Improve usage of kernel panic when pulling updates

davidegrohmann opened the pull request:
This PR makes sure that when applying transaction on slaves though the `TransactionResponseUnpacker` we kernel panic not only when we fail to append tx to the WAL but also when we fail in applying those to the store.  Moreover, we make sure that before starting to commit transaction in the `TransactionResponseUnpacker` we fail if the kernel is not healthy.

davidegrohmann committed:
Improve kernel panic when applying transactions to the store

Before the transaction commit process was in charge to raise a kernel
panic in case of failure in transaction application and since the
commit process is different between single and HA we had to duplicate
that logic in several places: TransactionRepresentationCommitProcess
and TransactionCommittingResponseUnpacker.

Now the responsibility for kernel panic is moved down into
TransactionRepresentationStoreApplier and IndexUpdatesValidator.
Moreover such a logic is not spread around multiple modules but it
lives only inside the kernel.

davidegrohmann committed:
Call kernelHealth.panic when failing to apply pulled transactions

Before we were making sure that kernelHealth.panic was called when
failing to append transaction to the log.  This could cause issues
since we cannot mark the state as failed and hence being able to check
point on a wrong state due to missing applied transactions.

davidegrohmann committed:
Check kernel health before applying transactions

In order to be on the safe side the
TransactionCommittingResponseUnpacker now checks kernel health before
committing the transactions received from the master.

MishaDemianenko commented:
@davidegrohmann can you please provide description how kernel panic will behave as result of this PR?

davidegrohmann commented:
@MishaDemianenko aren't commit messages enough?

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -581,7 +581,7 @@ public void transactionRecovered( long txId )
         {
             // Something unexpected happened during startup
-            msgLog.error(
+            msgLog.warn(
```
Why was this downgraded to a warning? Seems quite serious to me.

davidegrohmann replied:
Don't know, it is `warn` in 2.3 as well, so when backporting this change happened. I'll change this back to `error`.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/OnlineIndexUpdatesValidator.java:
```diff
@@ -55,11 +58,18 @@ public OnlineIndexUpdatesValidator( NeoStore neoStore, PropertyLoader propertyLo
+        {
+            transaction.accept( extractor );
+        }
+        catch ( IOException cause )
```
Is this really a panic? This is before we even start appending to log and committing, right?

davidegrohmann replied:
No I don't think it is, I'll remove that.

davidegrohmann replied:
Hmm, it looks like there is a reason. In both `RecoveryVisitor` and `TransactionCommittingResponseUnpacker` we check validity before applying to the store. Moreover, for the latter we actually append txs to the log first and  we validate them only before applying them on the store. So it looks like something is wrong here.

tinwelint commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpacker.java:
```diff
@@ -219,12 +222,6 @@ private void applyQueuedTransactions() throws IOException
             }
-            catch ( IOException e )
-            {
-                // Kernel panic is done on this level, i.e. append and apply doesn't do that themselves.
```
This panic might still be needed, because what if we get an OOM when we try to close the transactions (after appending)?

davidegrohmann replied:
Do you mean when calling `batchCloser`?

tinwelint replied:
Yup

davidegrohmann replied:
Right, I'll add a kernel panic there as well. Good catch.

davidegrohmann replied:
Fixed by calling `kernelHeath.panic(..)` in `BatchingTransactionRepresentationStoreApplier.closeBatch()`

MishaDemianenko commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpacker.java:
```diff
@@ -188,6 +188,9 @@ private void applyQueuedTransactions() throws IOException
             logRotation.rotateLogIfNeeded( LogAppendEvent.NULL );
 
+            // Check kernel health after log rotation
+            kernelHealth.assertHealthy( IOException.class );
```
Does it mean that as soon as we will detect panic here slave instance will be unusable till next election will be completed?

tinwelint replied:
Yes, manual intervention would be required. There is no panic handler... currently, that's part of another upcoming change.

MishaDemianenko replied:
Sure, can we here atm at least log - that we kind of dead and restart is required?

tinwelint replied:
@davidegrohmann @MishaDemianenko On that note: we log that inside `KernelHealth`. That message though is outdated and comes from earlier days where we had a TransactionManager class, refered to as `TM` in there. There are also other variable names refering to that time. We should rather fix that message and maybe make that note there, that the instance probably requires manual intervention and recovery in order to come back online.

davidegrohmann replied:
Alright I'll add that logging somehow. Though I rather prefer to perform the logging outside `KernelHealth`. So we can clean up `KernelHealth` later when designing a smarter kernel panic handler.

MishaDemianenko replied:
@davidegrohmann  @tinwelint i was looking into `panicMessage` and i would not treat that message as 'instance dead' or restart or we warn you! It's some problem, has something to do with  a tx.
As part of this pr we can just update that message to be more scary-realistic-descriptive, but in general i believe we need to provide more accurate useful specific messages at least and behaves in different way depending from exception encountered.

davidegrohmann committed:
Check validity of transactions before appending them to the log

When unpacking transactions we were checking their validity before
applying them to the store but after having appended them to the log.
This could cause problems since once the transaction is in the log is
always replayed during recovery, hence we might never recover when non
valid transactions are received.  This commit fixes that by checking
validity of transactions before appending them to the log.

That said, validate transactions is not an operations that should
cause a kernel panic, indeed if nothing is written in the log the
database is in a healthy state.  So when the validation step fails we
simply throw an exception.

MishaDemianenko commented on enterprise/com/src/main/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpacker.java:
```diff
@@ -188,6 +196,13 @@ private void applyQueuedTransactions() throws IOException
+            // Check kernel health after log rotation
+            if ( !kernelHealth.isHealthy() )
+            {
+                log.error( msg );
```
Lets log what actually was wrong in kernel? That should simplify investigation.

davidegrohmann replied:
The error is logged when we call `kernelHealth.panic(...)`, here we don't know what went wrong... we simply know that the kernel is not healthy.

MishaDemianenko replied:
we still have 'causeOfPanic' in kernelHealth so we can log actual problem.
And yes it's true that we log it initially, but i would assume as soon as we become unhealthy we will log plenty of messages into log file and it will be hard to find initial cause, if not impossible because we can rotate log file and delete it.
